nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/StandardLoadBalanceProtocol.java

```diff
@@ -0,0 +1,578 @@
+import static org.apache.nifi.controller.queue.clustered.protocol.LoadBalanceProtocolConstants.NO_MORE_FLOWFILES;
+import static org.apache.nifi.controller.queue.clustered.protocol.LoadBalanceProtocolConstants.REJECT_CHECKSUM;
+import static org.apache.nifi.controller.queue.clustered.protocol.LoadBalanceProtocolConstants.REQEUST_DIFFERENT_VERSION;
+import static org.apache.nifi.controller.queue.clustered.protocol.LoadBalanceProtocolConstants.VERSION_ACCEPTED;
+
+public class StandardLoadBalanceProtocol implements LoadBalanceProtocol {
+    private static final Logger logger = LoggerFactory.getLogger(StandardLoadBalanceProtocol.class);
+
+    private static final int SOCKET_CLOSED = -1;
+    private static final int NO_DATA_AVAILABLE = 0;
+
+    private final FlowFileRepository flowFileRepository;
+    private final ContentRepository contentRepository;
+    private final ProvenanceRepository provenanceRepository;
+    private final FlowController flowController;
+    private final LoadBalanceAuthorizer authorizer;
+
+    private final ThreadLocal<byte[]> dataBuffer = new ThreadLocal<>();
+    private final AtomicLong lineageStartIndex = new AtomicLong(0L);
+
+    public StandardLoadBalanceProtocol(final FlowFileRepository flowFileRepository, final ContentRepository contentRepository, final ProvenanceRepository provenanceRepository,
+                                       final FlowController flowController, final LoadBalanceAuthorizer authorizer) {
+        this.flowFileRepository = flowFileRepository;
+        this.contentRepository = contentRepository;
+        this.provenanceRepository = provenanceRepository;
+        this.flowController = flowController;
+        this.authorizer = authorizer;
+    }
+
+
+    @Override
+    public void receiveFlowFiles(final Socket socket) throws IOException {
+        final InputStream in = new BufferedInputStream(socket.getInputStream());
+        final OutputStream out = new BufferedOutputStream(socket.getOutputStream());
+
+        String peerDescription = socket.getInetAddress().toString();
+        if (socket instanceof SSLSocket) {
+            final SSLSession sslSession = ((SSLSocket) socket).getSession();
+
+            final Set<String> certIdentities;
+            try {
+                certIdentities = getCertificateIdentities(sslSession);
+
+                final String dn = CertificateUtils.extractPeerDNFromSSLSocket(socket);
+                peerDescription = CertificateUtils.extractUsername(dn);
+            } catch (final CertificateException e) {
+                throw new IOException("Failed to extract Client Certificate", e);
+            }
+
+            logger.debug("Connection received from peer {}. Will perform authorization against Client Identities '{}'",
+                peerDescription, certIdentities);
+
+            authorizer.authorize(certIdentities);
+            logger.debug("Client Identities {} are authorized to load balance data", certIdentities);
```

Comment:
Just a question, is there any reason for this authorization to be here? I was wondering if it can be done earlier at ConnectionLoadBalanceServer AcceptConnection.run or CommunicateAction.run.

Reply:
I don't think there's any particular reason that I put it here as opposed to the ConnectionLoadBalanceServer...